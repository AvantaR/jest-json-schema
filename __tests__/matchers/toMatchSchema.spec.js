const toMatchSchema = require('../../matchers/toMatchSchema');

expect.extend({ toMatchSchema });

describe('toMatchSchema', () => {
  let schema;

  beforeEach(() => {
    schema = {
      properties: {
        hello: { type: 'string' },
      },
      required: ['hello'],
    };
  });

  it('does not throw', () => {
    expect({ hello: 'world' }).toMatchSchema(schema);
    expect({ hello: 'a', world: 'b' }).toMatchSchema(schema);
    expect({}).not.toMatchSchema(schema);
    expect({ hello: 1 }).not.toMatchSchema(schema);
  });

  it('fails for wrong type', () => {
    const testObj = { hello: 1 };
    expect(() => expect(testObj).toMatchSchema(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('fails for missing required keys', () => {
    expect(() => expect({}).toMatchSchema(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('fails when pattern does not match', () => {
    schema.properties.hello.pattern = '[a-z]+';
    const testObj = { hello: '123' };
    expect(() => expect(testObj).toMatchSchema(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('fails when additional properties are found but forbidden', () => {
    schema.additionalProperties = false;
    const testObj = {
      hello: 'world',
      another: 'property',
    };
    expect(() => expect(testObj).toMatchSchema(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('includes the description in the error when provided', () => {
    const testObj = { hello: 1 };
    expect(() => expect(testObj).toMatchSchema(schema, 'en-US language pack'))
    .toThrowErrorMatchingSnapshot();
  });

  it('fails for matching schema when using .not', () => {
    const testObj = { hello: 'world' };
    expect(() => expect(testObj).not.toMatchSchema(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('does not crash on circular references', () => {
    const testObj = {};
    testObj.hello = testObj;
    expect(() => expect(testObj).toMatchSchema(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('assertion error matcherResult property contains matcher name and actual value', () => {
    schema.additionalProperties = false;
    const testObj = { another: 'property' };
    try {
      expect(testObj).toMatchSchema(schema);
    } catch (error) {
      expect(error.matcherResult).toEqual({
        actual: testObj,
        message: expect.any(Function),
        name: 'toMatchSchema',
        pass: false,
      });
    }
  });

  describe('custom formats', () => {
    describe('bcp47', () => {
      beforeEach(() => {
        schema = {
          properties: {
            locale: { type: 'string', format: 'bcp47' },
          },
        };
      });

      [
        'en-US',
        'nl-NL',
        'xx-XX',
      ].forEach((locale) => {
        it(`it matches ${locale}`, () => {
          expect({ locale }).toMatchSchema(schema);
        });
      });

      [
        'en-us',
        'EN-US',
        'en_US',
        'enus',
        '123',
      ].forEach((locale) => {
        it(`it does not match ${locale}`, () => {
          expect(() => expect({ locale }).toMatchSchema(schema))
            .toThrowErrorMatchingSnapshot();
        });
      });
    });
  });
});
