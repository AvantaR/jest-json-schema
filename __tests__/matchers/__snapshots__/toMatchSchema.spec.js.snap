// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchSchema custom formats bcp47 it does not match 123 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .locale should match format \\"bcp47\\"
[39m"
`;

exports[`toMatchSchema custom formats bcp47 it does not match EN-US 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .locale should match format \\"bcp47\\"
[39m"
`;

exports[`toMatchSchema custom formats bcp47 it does not match en_US 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .locale should match format \\"bcp47\\"
[39m"
`;

exports[`toMatchSchema custom formats bcp47 it does not match en-us 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .locale should match format \\"bcp47\\"
[39m"
`;

exports[`toMatchSchema custom formats bcp47 it does not match enus 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .locale should match format \\"bcp47\\"
[39m"
`;

exports[`toMatchSchema does not crash on circular references 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .hello should be string
[39m"
`;

exports[`toMatchSchema fails for matching schema when using .not 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toMatchSchema([22m[32mschema[39m[2m)[22m

Expected value not to match schema"
`;

exports[`toMatchSchema fails for missing required keys 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  should have required property 'hello'
[39m"
`;

exports[`toMatchSchema fails for wrong type 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .hello should be string
[39m"
`;

exports[`toMatchSchema fails when additional properties are found but forbidden 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  should NOT have additional properties, but found 'another'
[39m"
`;

exports[`toMatchSchema fails when pattern does not match 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  .hello should match pattern \\"[a-z]+\\"
[39m"
`;

exports[`toMatchSchema includes the description in the error when provided 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

en-US language pack
[31m  .hello should be string
[39m"
`;
